### PR TITLE
[full-ci][tests-only]Expand tests coverage related to user with different role

### PR DESCRIPTION
**Pull request by amrita-shrestha**
## Description
This PR expand tests coverage related to  the user with all role to create, edit, delete, get the user
As Viktor Scharf told `403 Forbidden vs 401 Unauthorized. I would be expect 403 code. But we definitely have chaos here and often for developers - 401 ok. in case non-existent - 404 or 403. looking at witch checking is first.`
So scenario like below has check for http status 401
```
- Non-admin user tries to delete his/her own account
- a normal user should not be able to edit another user's display name
- a normal user should not be able to reset the password of another user
- non-admin user tries to get the information of a user
- non-admin user tries to get all users
- non-admin user tries to get the group information of a user
- non-admin user tries to get users with a certain role
```

As Artur suggested running the authorization and non-existent related scenarios and adding them to the expected failure file

## Related Issue

- https://github.com/owncloud/ocis/issues/5565

## How Has This Been Tested?

- Locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/acceptance/features/apiGraph/deleteUser.feature [hunk 1]
```diff
@@ -55,22 +56,56 @@ Feature: delete user
...
+      | role        |
+      | Space Admin |
+      | User        |
+      | Guest       |
+
+
   Scenario: the admin user tries to delete a non-existent user
     Given the administrator has given "Alice" the role "Admin" using the settings api
-    When the user "Alice" deletes a user "nonExistentUser" using the Graph API
+    When the user "Alice" tries to delete a non-existing user using the Graph API
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/deleteUser.feature:73]:
```suggestion
    When the user "Alice" tries to delete a user "non-existing" using the Graph API
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 2]
```diff
@@ -421,6 +421,28 @@ public function removeUserFromGroup(string $groupId, string $userId, ?string $by
...
+				$this->featureContext->getStepLineRef(),
+				$credentials["username"],
+				$credentials["password"],
+				$userId
+			)
+		);
+	}
+
 	/**
 	 * @When /^the user "([^"]*)" deletes a user "([^"]*)" using the Graph API$/
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:447]:
```suggestion
	 * @When /^the user "([^"]*)" deletes a user "([^"]*)" using the Graph API$/
	 * @When /^the user "([^"]*)" tries to delete a user "([^"]*)" using the Graph API$/
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 3]
```diff
@@ -432,7 +454,22 @@ public function removeUserFromGroup(string $groupId, string $userId, ?string $by
 	 * @throws GuzzleException
 	 */
 	public function theUserDeletesAUserUsingTheGraphAPI(string $byUser, string $user): void {
-		$this->adminDeletesUserUsingTheGraphApi($user, $byUser);
+		$userId = $this->featureContext->getUserIdByUserName($user);
+		$this->adminDeletesUserByUserIdUsingTheGraphApi($userId, $byUser);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:458]: We can use username to delete user, so no need to get the user ID
```suggestion
		$this->adminDeletesUserUsingTheGraphApi($user, $byUser);
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 4]
```diff
@@ -432,7 +454,22 @@ public function removeUserFromGroup(string $groupId, string $userId, ?string $by
...
+	 *
+	 * @param string $byUser
+	 *
+	 * @return void
+	 * @throws GuzzleException
+	 * @throws Exception
+	 */
+	public function theUserTriesToDeleteNonExistingUser(string $byUser): void {
+		$userId = WebDavHelper::generateUUIDv4();
+		$this->adminDeletesUserByUserIdUsingTheGraphApi($userId, $byUser);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:472]:
```suggestion
```

File: tests/TestHelpers/GraphHelper.php [hunk 5]
```diff
@@ -285,6 +285,33 @@ public static function deleteUser(
...
+		$url = self::getFullUrl($baseUrl, 'users/' . $userId);
+
+		return HttpRequestHelper::delete(
+			$url,
+			$xRequestId,
+			$adminUser,
+			$adminPassword,
+		);
+	}
+
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:314]: IMO, not needed for this PR
```suggestion
```

**Review by saw-jan — COMMENTED** [review]
We can use username to delete user, so no need to get the user ID.
IMO, not needed for this PR

File: tests/TestHelpers/GraphHelper.php (hunk 5 above)
Comment (amrita-shrestha) [reply tests/TestHelpers/GraphHelper.php:314]: i asked Viktor which method should we use he told me to use `user id` to delete the user. So i implemented deletion of user using `user-id`

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 3 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/GraphContext.php:458]: @saw-jan answer of all your suggested changes related to this `i asked Viktor which method should we use he told me to use user id to delete the user. So i implemented deletion of user using user-id`

**Comment by amrita-shrestha** [comment]
> We can use username to delete user, so no need to get the user ID. IMO, not needed fo
i asked Viktor which method should we use he told me to use user id to delete the user. So i implemented deletion of user using user-id

File: tests/TestHelpers/GraphHelper.php (hunk 5 above)
Comment (saw-jan) [reply tests/TestHelpers/GraphHelper.php:314]: Was just trying to reduce the code changes if we can the existing helpers. But yeah, using ID seems to be the way to go as the clients use ids

**Comment by saw-jan** [comment]
^
Was just trying to reduce the code changes if we can the existing helpers. But yeah, using ID seems to be the way to go as the clients use ids

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 6]
```diff
@@ -421,6 +421,28 @@ public function removeUserFromGroup(string $groupId, string $userId, ?string $by
...
+			GraphHelper::deleteUserByUserId(
+				$this->featureContext->getBaseUrl(),
+				$this->featureContext->getStepLineRef(),
+				$credentials["username"],
+				$credentials["password"],
+				$userId
+			)
+		);
+	}
+
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:445]:
```suggestion
```

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 4 above)
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:472]:
```suggestion
	public function theUserTriesToDeleteNonExistingUser(string $byUser): void {
		$userId = WebDavHelper::generateUUIDv4();
		$credentials = $this->getAdminOrUserCredentials($byUser);
		$this->featureContext->setResponse(
			GraphHelper::deleteUserByUserId(
				$this->featureContext->getBaseUrl(),
				$this->featureContext->getStepLineRef(),
				$credentials["username"],
				$credentials["password"],
				$userId
			)
		);
```

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 3 above)
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:458]:
```suggestion
		$userId = $this->featureContext->getUserIdByUserName($user);
		$credentials = $this->getAdminOrUserCredentials($byUser);
		$this->featureContext->setResponse(
			GraphHelper::deleteUserByUserId(
				$this->featureContext->getBaseUrl(),
				$this->featureContext->getStepLineRef(),
				$credentials["username"],
				$credentials["password"],
				$userId
			)
		);
```

**Review by saw-jan — COMMENTED** [review]
use `setResponse()` within the `When` implementation only.

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 3 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/GraphContext.php:458]: why we have to duplicate codes like this? We can simply return response and setResponse in When step

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 4 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/GraphContext.php:472]: why we have to duplicate codes like this? We can simply return response and setResponse in When step

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 6 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/GraphContext.php:445]: this help us to implement DRY concept

**Comment by amrita-shrestha** [comment]
> use `setResponse()` within the `When` implementation only.

@saw-jan if we have to setResponse() in When step then we can do that by returning response from function why to remove function we help to implement DRY concept?

**Comment by saw-jan** [comment]
> @saw-jan if we have to setResponse() in When step then we can do that by returning response from function why to remove function which help to implement DRY concept?

Yeah, my only suggestion is to use `setResponse()` within When. Ignore others

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 7]
```diff
@@ -421,6 +421,26 @@ public function removeUserFromGroup(string $groupId, string $userId, ?string $by
...
+	/**
+	 * sends a request to delete a user with the help of userID using the Graph API
+	 *
+	 * @param string $userId
+	 * @param string $byUser
+	 *
+	 * @return ResponseInterface
+	 * @throws GuzzleException
+	 */
+	public function adminDeletesUserByUserIdUsingTheGraphApi(string $userId, string $byUser): ResponseInterface {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:433]:
```suggestion
	public function deleteUserByUserIdUsingTheGraphApi(string $userId, string $byUser): ResponseInterface {
```

File: tests/acceptance/features/apiGraph/editUser.feature [hunk 8]
```diff
@@ -156,21 +180,32 @@ Feature: edit user
...
 
   Scenario Outline: a normal user should not be able to reset the password of another user
-    Given the administrator has given "Brian" the role "<role>" using the settings api
+    Given the administrator has given "Brian" the role "<userRole>" using the settings api
     And the user "Alice" has created a new user using the Graph API with the following settings:
       | userName    | Carol             |
       | displayName | Carol King        |
       | email       | carol@example.com |
       | password    | 1234              |
+    Given the administrator has given "Carol" the role "<role>" using the settings api
```
Comment (PrajwolAmatya) [inline tests/acceptance/features/apiGraph/editUser.feature:189]:
```suggestion
    And the administrator has given "Carol" the role "<role>" using the settings api
```

File: tests/acceptance/features/apiGraph/deleteUser.feature [hunk 9]
```diff
@@ -55,22 +56,56 @@ Feature: delete user
...
     Then the HTTP status code should be "404"
 
 
+  Scenario Outline: Non-admin user tries to delete a non-existent user
+    Given the administrator has given "Alice" the role "<role>" using the settings api
+    When the user "Alice" tries to delete a non-existing user using the Graph API
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/deleteUser.feature:79]: just that we are using `nonexistent` in other places so good to maintain it
```suggestion
    When the user "Alice" tries to delete a nonexistent user using the Graph API
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 10]
```diff
@@ -20,10 +20,26 @@ Feature: get users
       | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    | true           |
 
 
-  Scenario: non-admin user tries to get the information of a user
+  Scenario Outline: non-admin user tries to get the information of a user
+    Given the administrator has given "Alice" the role "<role>" using the settings api
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:24]:
```suggestion
    Given the administrator has given "Alice" the role "<role>" using the settings api
```
Step duplication with L#12

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 11]
```diff
@@ -20,10 +20,26 @@ Feature: get users
       | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    | true           |
 
 
-  Scenario: non-admin user tries to get the information of a user
+  Scenario Outline: non-admin user tries to get the information of a user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:23]: tag the issue also, here and in other scenarios as well
```suggestion
  @issue-5125
  Scenario Outline: non-admin user tries to get the information of a user
```

File: tests/acceptance/expected-failures-localAPI-on-OCIS-storage.md [hunk 12]
```diff
@@ -92,18 +92,45 @@ The expected failures in this file are from features in the owncloud/ocis repo.
...
+#### [Space admin should not be able to change the user quota](https://github.com/owncloud/ocis/issues/5475)
 - [apiSpaces/spaceManagement.feature:149](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/spaceManagement.feature#L149)
 
 #### [Normal user can get expanded members information of a group](https://github.com/owncloud/ocis/issues/5604)
 - [apiGraph/getGroup.feature:100](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getGroup.feature#L100)
 
+#### [Non-admin user tries to delete non-existing result 401](https://github.com/owncloud/ocis/issues/5738)
+- [apiGraph/deleteUser.feature:83](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/deleteUser.feature#L83)
+- [apiGraph/deleteUser.feature:84](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/deleteUser.feature#L84)
+- [apiGraph/deleteUser.feature:85](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/deleteUser.feature#L85)
```
Comment (saw-jan) [inline tests/acceptance/expected-failures-localAPI-on-OCIS-storage.md:132]: Not sure whether to expect this or not. IMO, non-admin user getting `401` while trying to delete non-existing user should be fine.
Also see https://github.com/owncloud/ocis/issues/5738#issuecomment-1471501646

CC @SwikritiT @SagarGi @grgprarup @PrajwolAmatya @individual-it

**Review by saw-jan — COMMENTED** [review]
LGTM

**Review by SagarGi — COMMENTED** [review]
LGTM :+1:

File: tests/acceptance/expected-failures-localAPI-on-OCIS-storage.md (hunk 12 above)
Comment (individual-it) [reply tests/acceptance/expected-failures-localAPI-on-OCIS-storage.md:132]: I think its a bug, but likely low prio. See my comment in the issue

**Review by SwikritiT — APPROVED** [review]
LGTM

**Review by saw-jan — APPROVED** [review]

**Review by ScharfViktor — APPROVED** [review]

**Comment by SwikritiT** [comment]
@amrita-shrestha just a reminder that this needs to be backported to stable-2.0 (with skip on stable tag probably)?
# TODO
- [ ] Backport to stable branch

**Pull request merged**